src/panoptoindexconnector/implementations/graph_implementation.py

```diff
@@ -0,0 +1,569 @@
+    elif response.status_code == 404:
+        LOG.info(f"Video ({video_id}) not found to be delete from target!")
+
+
+#
+# Initialize and teardown steps here
+#
+
+def initialize(config):
+    """
+    Create connection and register schema if not exists
+    """
+
+    try:
+        ensure_connection_availability(config)
+    except Exception as ex:
+        LOG.error(f'Error occurred while initializing graph connector!. Error: {ex}')
+        raise
+
+
+def teardown(config):
+    """
+    Delete token cache file
+    """
+
+    if os.path.exists(TOKEN_CACHE):
+        os.remove(TOKEN_CACHE)
+
+
+#########################################################################
+#
+# Local helpers
+#
+#########################################################################
+
+
+def set_main_fields_to_new_object(video_content):
+    """
+    Set id and content properties to new object
+    """
+
+    return {
+        "id": video_content["Id"],
+        "content": {
+            "type": "text",
+            "value": "{0}{1}".format(video_content["Title"], " " + video_content["Summary"] if video_content["Summary"] else "")
+        }
+    }
+
+
+def set_properties(field_mapping, panopto_content, target_content):
+    """
+    Set properties
+    """
+
+    # Set properties from Info yaml
+    target_content['properties'] = {
+        field:  panopto_content['VideoContent'][key]
+        for key, field in field_mapping['Info'].items()
+        if panopto_content['VideoContent'][key]
+    }
+
+    # Set properties from Metadata yaml
+    target_content["properties"].update({
+        field: panopto_content['VideoContent'][key]
+        for key, field in field_mapping['Metadata'].items()
+        if panopto_content['VideoContent'][key]
+    })
+
+
+def set_principals(config, panopto_content, target_content):
+    """
+    Set principals - Everyone (in tenant), User or Group.
+    If none of permissions are set, target_content will be set to None and skipped for sync
+    """
+
+    if not config.skip_permissions:
+        if is_public_or_all_users_principals(panopto_content):
+            set_principals_to_all(config, target_content)
+        elif is_user_principals(panopto_content):
+            set_principals_to_user(config, panopto_content, target_content)
+    else:
+        set_principals_to_all(config, target_content)
+
+    if not target_content["acl"]:
+        target_content = None
+
+        LOG.warn("Target content will be skipped to push to target since none of permissions have applied!")
+
+
+def get_unique_principals(panopto_content):
+    """
+    Get unique principals to avoid duplicate permissions on synced item
+    """
+
+    unique_content_principals = []
+
+    for principal in panopto_content['VideoContent']['Principals']:
+        if principal not in unique_content_principals:
+            unique_content_principals.append(principal)
+
+    return unique_content_principals
```

Comment:
nit: could just use a set
```suggestion
    unique_content_principals = set(panopto_content['VideoContent']['Principals'])
    return list(unique_content_principals )
```

Reply:
I am not able to use `set` on principals since principals is dictionary = unhashable type.